### PR TITLE
Add min_pool_size, Add default value of should_shuffle

### PR DESCRIPTION
**Pull request by reyoung**
* min_pool_size would be infinite by default.
  * add unittest for min_pool_size
* Fix bug in can_over_batch_size
  * add unittest for can_over_batch_size
* Add DEFINE_PROVIDER_EX
* Add default value of should_shuffle
  * When training, the default value of should_shuffle is True.
  * When testing, the default value of should_shuffle is False.
  * User a set a provider should_shuffle or not by pass it to `@provider`
  * should_shuffle can handle a list of value, not just boolean
* Add input order mapping by using name
  * Add unittest
* Add check to check input format.
  * Default is close for speed reason.
  * User could stop train when check error, or continue train without
    this train sample.
* use deque instead of vector in generators pool, make erase
  generator faster.
* Add chinese/english documentation
* Make should shuffle = false in unittest
* Add python files to depends.

### CODE REVIEW

File: paddle/gserver/dataproviders/DataProvider.cpp [hunk 1]
```diff
@@ -149,9 +149,13 @@ void DoubleBuffer::startAsyncLoad() {
...
-ClassRegistrar<DataProvider, DataConfig, bool> DataProvider::registrar_;
-DataProvider* DataProvider::create(const DataConfig& config, bool useGpu) {
-  return registrar_.createByType(config.type(), config, useGpu);
+ClassRegistrar<DataProvider, DataConfig, ModelConfig, bool>
+DataProvider::registrar_;
+
+DataProvider* DataProvider::create(const DataConfig& config,
+                                   const ModelConfig& modelConfig,
+                                   bool useGpu) {
+  return registrar_.createByType(config.type(), config, modelConfig, useGpu);
```
Comment (reyoung) [inline paddle/gserver/dataproviders/DataProvider.cpp]: Add ModelConfig in `DataProvider::create` to get input layer order

File: paddle/gserver/dataproviders/DataProvider.h [hunk 2]
```diff
@@ -39,15 +39,24 @@ limitations under the License. */
...
-  })
+#define REGISTER_DATA_PROVIDER(__type_name, __class_name)\
+  static InitFunction __reg_type_##__type_name([]() {\
+  DataProvider::registrar_.registerClass(\
+  #__type_name, \
+  [](DataConfig conf, ModelConfig, bool useGpu) -> DataProvider* { \
+    DataProvider* dp = new __class_name (conf, useGpu);\
+    return dp;\
+  });\
+})
```
Comment (reyoung) [inline paddle/gserver/dataproviders/DataProvider.h]: REGISTER_DATA_PROVIDER for DataProvider not use ModelConfig

File: paddle/gserver/dataproviders/DataProvider.h [hunk 3]
```diff
@@ -39,15 +39,24 @@ limitations under the License. */
...
+
+#define REGISTER_DATA_PROVIDER_EX(__type_name, __class_name)            \
+  static InitFunction __reg_type_##__type_name([] {                     \
+  DataProvider::registrar_.registerClass<__class_name>(#__type_name);   \
+})
```
Comment (reyoung) [inline paddle/gserver/dataproviders/DataProvider.h]: REGISTER_DATA_PROVIDER_EX for DataProvider use ModelConfig

File: paddle/gserver/dataproviders/PyDataProvider2.cpp [hunk 4]
```diff
@@ -24,6 +24,27 @@ limitations under the License. */
 
 namespace paddle {
 
+namespace unittest {
```
Comment (reyoung) [inline paddle/gserver/dataproviders/PyDataProvider2.cpp]: used only for unittest

File: paddle/gserver/dataproviders/PyDataProvider2.cpp [hunk 5]
```diff
@@ -521,8 +565,17 @@ class PyDataProvider2 : public DataProvider {
           calcBatchSize.getArgs().set(0, data.back());
           PyObjectPtr customBatchSize(calcBatchSize());
           bool ok;
-          bsize += py::castInt<size_t>(customBatchSize.get(), &ok);
+          size_t tmp = py::castInt<size_t>(customBatchSize.get(), &ok);
           CHECK(ok) << "calc_batch_size must return int";
+
+          if (bsize + tmp > size && !canOverBatchSize_) {
+            // Put data back.
```
Comment (reyoung) [inline paddle/gserver/dataproviders/PyDataProvider2.cpp]: Put data back when larger then batch size

File: paddle/gserver/dataproviders/DataProvider.h [hunk 6]
```diff
@@ -39,15 +39,24 @@ limitations under the License. */
...
+#define REGISTER_DATA_PROVIDER(__type_name, __class_name)\
+  static InitFunction __reg_type_##__type_name([]() {\
+  DataProvider::registrar_.registerClass(\
+  #__type_name, \
+  [](DataConfig conf, ModelConfig, bool useGpu) -> DataProvider* { \
+    DataProvider* dp = new __class_name (conf, useGpu);\
+    return dp;\
+  });\
+})
+
```
Comment (emailweixu) [inline paddle/gserver/dataproviders/DataProvider.h]: comment

File: paddle/gserver/dataproviders/DataProvider.h [hunk 7]
```diff
@@ -39,15 +39,24 @@ limitations under the License. */
...
-
 /**
  * @def REGISTER_DATA_PROVIDER
  * @brief Macro for registering a data provider
  */
-#define REGISTER_DATA_PROVIDER(__type_name, __class_name)               \
-  static InitFunction __reg_type_##__type_name([]() {                   \
-    DataProvider::registrar_.registerClass<__class_name>(#__type_name); \
-  })
+#define REGISTER_DATA_PROVIDER(__type_name, __class_name)\
```
Comment (emailweixu) [inline paddle/gserver/dataproviders/DataProvider.h]: comment

File: python/paddle/trainer/PyDataProvider2.py [hunk 8]
```diff
@@ -102,13 +117,97 @@ def __init__(self, generator):
...
+        self.input_types = input_types
+        self.check_fail_continue = check_fail_continue
+        self.logger = logger
+
+    def __call__(self, obj, filename):
+        for items in self.generator(obj, filename):
+            try:
+                assert len(items) == len(self.input_types)
+                assert len(filter(lambda x: x is None, items)) == 0
+                for item, input_type in zip(items, self.input_types):
```
Comment (emailweixu) [inline python/paddle/trainer/PyDataProvider2.py]: zip => itertools.izip

File: paddle/gserver/tests/test_PyDataProvider2.cpp [hunk 9]
```diff
@@ -235,6 +247,107 @@ TEST(PyDataProvider2, index_sub_seq) {
...
+  config.set_files(FLAGS_train_list.c_str());
+  config.set_load_data_module("test_PyDataProvider2");
+  config.set_load_data_object("test_check");
+  config.set_load_data_args("");
+  paddle::DataBatch batch;
+  std::unique_ptr<paddle::DataProvider> provider(
+  paddle::DataProvider::create(config, false));
+  provider->reset();
+  while (true) {
+    size_t realBatchSize = provider->getNextBatchInternal(100, &batch);
```
Comment (emailweixu) [inline paddle/gserver/tests/test_PyDataProvider2.cpp]: should check the values are in [0, 10) here

File: python/paddle/trainer/PyDataProvider2.py [hunk 10]
```diff
@@ -127,10 +226,27 @@ def process(settings, file_name):
...
 
+    :param check: Check the yield data format is as same as input_types. Enable
+                  this will make data provide process slow but it is very useful
+                  for debug. Default is disabled.
+    :type check: bool
+    :param check_fail_continue: Continue train or not when check failed. Just
+                                drop the wrong format data when it is True. Has
+                                no effect when check set to False.
+    :type check_fail_continue: bool
+    :param use_dynamic_order: Make provider can yield a dictionary object. Which
```
Comment (emailweixu) [inline python/paddle/trainer/PyDataProvider2.py]: Make provider can yield a dictionary object, whose

**Comment by emailweixu** [comment]
Also please update the data provider documentation

**Comment by reyoung** [comment]
@emailweixu  Update codes. Add chinese docs. Engligh document will be added asap.

**Comment by emailweixu** [comment]
Need to fix test

**Comment by reyoung** [comment]
@emailweixu The unittest error before, is because we didn't disable shuffle when unittest and this patch set the `min_pool_size` to unlimited. It makes data shuffle correct, and influences the unittest before.

File: doc/ui/data_provider/pydataprovider2.rst [hunk 11]
```diff
@@ -155,23 +168,7 @@ Reference
...
-* can_over_batch_size defines whether PaddlePaddle can store little more
-  samples than pool_size. It is better to set True to avoid some deadlocks.
-* calc_batch_size is a function define how to calculate batch size. This is
-  usefull in sequential model, that defines batch size is counted upon sequence
-  or token. By default, each sample or sequence counts to 1 when calculating
-  batch size.
-* cache is a data cache strategy, see `cache`_.
-* Init_hook function is invoked once the data provider is initialized,
-  see `init_hook`_.
+.. autofunction:: paddle.trainer.PyDataProvider2.provider
```
Comment (reyoung) [inline doc/ui/data_provider/pydataprovider2.rst]: Here, we use paddle.trainer.PyDataProvider2.provider's comments as documentation.

**Pull request merged**